### ISSUE TITLE
[Bug]: Performance Issue: Significant Time Difference when Plotting Time Series Data with and without Timezones

### ISSUE THREAD

**Issue by HarryCollins2** [matplotlib/matplotlib#27717]
### Bug summary

Hello. When plotting time series data using Matplotlib and Pandas, I have observed a significant difference in execution times between plots with and without timezones. 

**Code and Observation:**

I have provided a code snippet below for reference:

### Code for reproduction

```python
import matplotlib.pyplot as plt

n = 1000000

vertical = [i for i in range(n)]

# Prepare the data
dates_no_tz = pd.date_range('2019-01-01', periods=n, freq='T')
dates_with_tz = pd.date_range('2019-01-01', periods=n, freq='T', tz='UTC')

# Plot the time series without timezone
%time plt.plot(dates_no_tz, vertical)

# Plot the time series with timezone
%time plt.plot(dates_with_tz, vertical)
```

### Actual outcome

CPU times: user 50.6 ms, sys: 19.1 ms, total: 69.7 ms
Wall time: 71.2 ms
CPU times: user 3.44 s, sys: 192 ms, total: 3.63 s
Wall time: 3.69 s

### Expected outcome

Expect a similar execution time

### Matplotlib Version

3.8.0

### Matplotlib Backend

module://matplotlib_inline.backend_inline

### Python version

Python 3.9.18

### Installation

pip

**Comment by HarryCollins2** [comment]
https://github.com/mwaskom/seaborn/issues/3627#issuecomment-1916687929, I want to express my appreciation for [mwaskom](https://github.com/mwaskom) for looking into this. It appears that the lag is indeed related to `plt.gca().xaxis.convert_units(df_with_tz["Time"])`

**Comment by thuiop** [comment]
I looked a bit into it and I am afraid there is not much to do. Much of the time is spent in a list comprehension, which is basically calling `datetime.astimezone` on each element from the input ; the remainder is spent converting to the `numpy.datetime64` format. Ideally, you would take care of the timezone stuff before plotting, in which case it skips this part and is much faster. Something like `np.arange(np.datetime64('2019-01-01'), np.datetime64('2019-01-01') + n*np.timedelta64(1,"m"), np.timedelta64(1,"m"))` would do the trick in the example case I think, although I suppose the actual case is more complex. In any case, this boils down to the time needed to make the timezone conversions so I do not think there is a way around this.
![Screenshot from 2024-01-30 14-16-21](https://github.com/matplotlib/matplotlib/assets/1338337/ea1c07d5-e12a-4cc4-be82-1ce35e2e9c6e)

**Comment by dstansby** [comment]
> Much of the time is spent in a list comprehension, which is basically calling datetime.astimezone on each element from the input

Is this happening in pandas or Matplotlib?

**Comment by thuiop** [comment]
> Is this happening in pandas or Matplotlib?

Matplotlib. More specifically, there : https://github.com/matplotlib/matplotlib/blob/41d4149f84b40d3a90a70f81b940985b8485f04a/lib/matplotlib/dates.py#L405-L454 
The offending part is in the `if` block starting on line 439 (which gets skipped in the example with no timezone, and is executed when there is one). The thing is, this block is supposed to (also) work with regular `datetime` from the standard library ; it conveniently works with pandas as pandas uses the same API. So unless you are willing to add pandas specific code, I do not think there is a faster way to do this.

**Comment by rcomer** [comment]
Thanks for digging into this @thuiop!  Do you know if Matplotlib's `date2num` gets called at all for the "no timezone" case?  Noting that Matplotlib ultimately needs a sequence of floats and Pandas presumably starts out with a sequence of ints/floats, my suspicion is that in the "no timezone" case we get the original numbers directly without ever converting to datetimes.  I have checked that

* Just plotting `vertical` against itself takes a similar amount of time as the "no timezone" case.
* If I generate a sequence of datetimes with
   `dates_no_tz = [datetime.datetime(2019, 1, 1) + datetime.timedelta(minutes=i) for i in range(n)]`
  it takes about 25x longer to plot against that than against `dates_no_tz` created with pandas.

**Comment by thuiop** [comment]
Yes, it is indeed called in the "no timezone" case ; however it is very fast in that case, as the underlying numpy array from pandas already has a `np.datetime64`-like dtype, which allows it to skip the (slow) timezone conversion. The thing is, with the current implementation, you first iterate over all the elements to convert the timezone and switch it to a naive datetime, and then recast it as a numpy array (which is also not free). Unless pandas-specific code is introduced, there is not a real way around this, since this needs to work for a list of `datetime` as in your second example ; I am pretty sure there is no fast way to convert from `datetime` to `np.datetime64`.

**Comment by jklymak** [comment]
For added evidence, this is also slow with pure `datetime`.  I don't know of any way to convert a list of datetimes faster than the list comprehension.  Could possibly modify in place?  

```python
import numpy as np
import matplotlib.pyplot as plt
import datetime
import pytz
import time

# Prepare the data
N = 1_000_000
vertical = [i for i in range(N)]
dates_no_tz = [datetime.datetime(2024, 1, 1) + datetime.timedelta(seconds=x) for x in range(N)]

# Plot the time series without timezone
start = time.time()
plt.plot(dates_no_tz, vertical)
print('Time no tz ', time.time()-start)
plt.show()

dates_w_tz = [pytz.timezone('Europe/Berlin').localize(datetime.datetime(2024, 1, 1)) + datetime.timedelta(seconds=x) for x in range(N)]

# Plot the time series with timezone
start = time.time()
plt.plot(dates_w_tz, vertical)
print('Time w timezone', time.time()-start)
plt.show()
```

```
Time no tz  1.9029
Time w timezone 3.3815
```

**Comment by QuLogic** [comment]
We can save a little here with `np.vectorize`, though I can't say whether it's really worth it.

```python
import datetime
import timeit
import pytz

# Prepare the data
N = 1_000_000
vertical = [i for i in range(N)]
dates_no_tz = [datetime.datetime(2024, 1, 1) + datetime.timedelta(seconds=x) for x in range(N)]

# Plot the time series without timezone
print('Time no tz ',
      timeit.timeit('ax.plot(dates_no_tz, vertical)',
                    setup='import matplotlib.pyplot as plt; plt.close("all"); _, ax = plt.subplots()',
                    globals=globals(), number=10) / 10)


dates_w_tz = [pytz.timezone('Europe/Berlin').localize(datetime.datetime(2024, 1, 1)) + datetime.timedelta(seconds=x) for x in range(N)]

# Plot the time series with timezone
print('Time w timezone',
      timeit.timeit('ax.plot(dates_w_tz, vertical)',
                    setup='import matplotlib.pyplot as plt; plt.close("all"); _, ax = plt.subplots()',
                    globals=globals(), number=10) / 10)
```
before:
```
Time no tz  2.1389926880015993
Time w timezone 4.112374102405738
```

and after this patch:
```diff
diff --git a/lib/matplotlib/dates.py b/lib/matplotlib/dates.py
index c12d9f31ba..25a7ddc414 100644
--- a/lib/matplotlib/dates.py
+++ b/lib/matplotlib/dates.py
@@ -374,6 +374,9 @@ def _from_ordinalf(x, tz=None):
 _from_ordinalf_np_vectorized = np.vectorize(_from_ordinalf, otypes="O")
 # a version of dateutil.parser.parse that can operate on numpy arrays
 _dateutil_parser_parse_np_vectorized = np.vectorize(dateutil.parser.parse)
+_make_tz_naive_vectorized = np.vectorize(
+    lambda dt: dt.astimezone(UTC).replace(tzinfo=None),
+    otypes="O")


 def datestr2num(d, default=None):
@@ -444,8 +447,7 @@ def date2num(d):
         tzi = getattr(d[0], 'tzinfo', None)
         if tzi is not None:
             # make datetime naive:
-            d = [dt.astimezone(UTC).replace(tzinfo=None) for dt in d]
-            d = np.asarray(d)
+            d = _make_tz_naive_vectorized(d)
         d = d.astype('datetime64[us]')

     d = np.ma.masked_array(d, mask=mask) if masked else d
```
is now:
```
Time no tz  2.122314026101958
Time w timezone 3.574551021598745
```

For Pandas, it's
```
In [6]: %time plt.plot(dates_with_tz, vertical)
CPU times: user 8.35 s, sys: 92.6 ms, total: 8.44 s
Wall time: 8.46 s
```
vs
```
In [6]: %time plt.plot(dates_with_tz, vertical)
CPU times: user 7.73 s, sys: 165 ms, total: 7.89 s
Wall time: 7.92 s
```
